platform-sdk/docs/proposals/TSS-Block-Signing/tss-block-signing.md

```diff
@@ -0,0 +1,309 @@
+captures the accumulation of share signatures after consensus. If we can tolerate the possibility of failing to
+construct a ledger signature, we can avoid the introduction of state and process the share signature transactions
+pre-consensus. This allows us to construct signatures faster during healthy execution.
+
+### Design Decisions
+
+No Guaranteed Ledger Signature:
+
+- `TssShareSignatureTransaction` are handled pre-consensus in the pre-handle phase of transaction handling
+  - this is to allow the creation of signatures as fast as possible.
+- The `TssShareSignatureTransaction` are not stored in the state.
+
+This decision (coming from the Block Stream design) is to have faster signatures handled in pre-consensus and
+cover any missed signatures on blocks with Merkle Proofs using signatures from later blocks.
+
+#### Alternatives Considered
+
+Guaranteed Ledger Signature Generation:
+
+- requires storing share signatures in the state.
+- requires processing share signatures after consensus, introducing a delay in block signing.
+
+This was not selected for the following reasons:
+
+1. block proofs do not need to be identical between nodes.
+2. the probability of failure to sign a block is low.
+3. speed in signing a block matters since blocks are not published without signatures once they are signable.
+
+## Changes
+
+The changes are presented in the following order:
+
+1. Public API
+2. Component Architecture
+3. Core Behaviors
+4. Configuration
+5. Metrics
+6. Performance
+
+### Public API
+
+#### Services
+
+The `TssBaseService` is extended with the following API:
+
+```Java
+/**
+ * The TssBaseService will attempt to generate TSS key material for any set candidate roster, giving it a ledger id and
+ * the ability to generate ledger signatures that can be verified by the ledger id.  Once the candidate roster has
+ * received its full TSS key material, it can be made available for adoption by the platform.
+ * </p>
+ * The TssBaseService will also attempt to generate ledger signatures by aggregating share signatures produced by
+ * calling {@link #requestLedgerSignature(byte[])}.
+ *
+ */
+public interface TssBaseService {
+
+    ...
+
+    /**
+     * Requests a ledger signature on a message hash.  The ledger signature is computed asynchronously and returned
+     * to all consumers that have been registered through {@link #registerLedgerSignatureConsumer(Consumer)}.
+     *
+     * @param messageHash The hash of the message to be signed by the ledger.
+     */
+    void requestLedgerSignature(@Nonnull byte[] messageHash);
+
+    /**
+     * Registers a consumer of pairs where the first element is the message hash and the second element is the
+     * ledger signature on the message hash.
+     *
+     * @param consumer the consumer of ledger signatures on message hashes.
+     */
+    void registerLedgerSignatureConsumer(@Nonnull final BiConsumer<byte[], PairingSignature> consumer);
```

Comment:
For completeness I suggest a,
```
    /**
     * Unregisters a consumer of pairs where the first element is the message hash and the second element is the
     * ledger signature on the message hash.
     *
     * @param consumer the consumer of ledger signatures on message hashes to unregister
     */
    void unregisterLedgerSignatureConsumer(@Nonnull final BiConsumer<byte[], PairingSignature> consumer);
```

Reply:
I went ahead and added it since I was editing the proposal.